### DISCUSSION TITLE
"Status 500: Unexpected Problem" when accessing grid's forecast

### DISCUSSION THREAD

**Discussion by ChaseBosman** [weather-gov/api#241] [Q&A]
**Describe the bug**
Occasionally when accessing a grid's weather forecast, a status 500 error is returned. The error message indicates that an unexpected problem has occurred.

**To Reproduce**
The link currently providing this error is: https://api.weather.gov/gridpoints/MTR/114,128

**If possible, also include the correlationId from the response**
1dc193d9

**Expected behavior**
The expected JSON response for a grid's forecast.

**Screenshots**
![image](https://user-images.githubusercontent.com/54570221/106550866-daadeb80-64c8-11eb-9e3b-b4da4aa1b8b9.png)

**Additional Context**
I have noticed that this is a rather intermittent issue.

Thanks,
Chase

**Comment by iampeterdaniels** [comment]
I've noticed this, too. I added retry looping with exponential backoff to my code. Usually the second or third GET will succeed.

**Comment by ChaseBosman** [comment]
It's interesting...revisiting this problem 12 days later, the same grid is still producing the error.

**Comment by whoinow** [comment]
I am also getting this problem even with 3 retries 1s apart. I am going to try backing off my retries, but I doubt that will help. I will say I am hitting a different endpoint though: https://api.weather.gov/gridpoints/MTR/83,153/forecast/hourly

**Comment by scadergit** [comment]
Status 500 errors are operational outages and can be reported to the helpdesk for a more immediate response. The original request has been purged from the logs, so I can no longer determinen the cause. If anyone can provide a more recent correlation id then I will look into it.

**Reply by strefethen** [reply]
Here's one that's failing now (Nov5, 11:24am PST)

https://api.weather.gov/gridpoints/MTR/110,118/forecast

I too see quite a few 500 errors on a daily basis.

**Comment by whoinow** [comment]
Unfortunately my logging doesn't have a correlationId. I make my requests at 12am PST. Are their planned outages at certain times? I looked [here](https://www.weather.gov/notification/) and didn't see anything.

**Comment by ChaseBosman** [comment]
> Status 500 errors are operational outages and can be reported to the helpdesk for a more immediate response. The original request has been purged from the logs, so I can no longer determinen the cause. If anyone can provide a more recent correlation id then I will look into it. 

Here is a recently generated correlation ID for the issue: 4a3c06e7

The link in the original post for this issue actually always generates this error for me, which is interesting seeing that the grid is a fairly large city in California.

**Comment by gerrypez** [comment]
Ex. 500 status errors (unexpected problem), not intermittent ... the correlation ID is the first line in the 500 error response.  

https://api.weather.gov/gridpoints/MTR/106,127/forecast/hourly
https://api.weather.gov/gridpoints/MTR/69,161/forecast/hourly
https://api.weather.gov/gridpoints/MTR/104,152/forecast/hourly
https://api.weather.gov/gridpoints/MTR/68,163/forecast/hourly

**Comment by scadergit** [comment]
A quick glance at the correlation ids indicate an issue with the service that translates lat/lon to grids. Support is typically able to address this issue. As always, please send all 500 errors to NCO support.

**Comment by yoonsoo-park** [comment]
Still getting 500 frequently. The interesting point is when I access directly to https://api.weather.gov/gridpoints/ILM/94,68/forecast, it works but when I request from my application, it returns 500. 
`{'correlationId': 'f46c1ec', 'title': 'Unexpected Problem', 'type': 'https://api.weather.gov/problems/UnexpectedProblem', 'status': 500, 'detail': 'An unexpected problem has occurred.', 'instance': 'https://api.weather.gov/requests/f46c1ec'}`
`{'correlationId': 'f4390c7', 'title': 'Unexpected Problem', 'type': 'https://api.weather.gov/problems/UnexpectedProblem', 'status': 500, 'detail': 'An unexpected problem has occurred.', 'instance': 'https://api.weather.gov/requests/f4390c7'}`

**Reply by strefethen** [reply]
I'm still seeing these as well and I suspect it's likely the issue @scadergit referred to last Nov. To me. the issue "feels like" a caching problem and if I browse to the link after the error it often will return the proper (and now cached) result. I could be totally wrong here but this is the sense that I get that the lat/long is perhaps queued and in between a 500 is returned.

**Reply by scadergit** [reply]
I confirmed it's the same issue: an upstream service that is going sideways. NCO support knows how to address the issue until the service can be replaced.

**Comment by dwhitemv25** [comment]
What threshold should we be reporting /gridpoints issues to NCO, being that its a known long-term issue? This is the counts of 500 errors for a call to `https://api.weather.gov/gridpoints/PQR/78,49/forecast` since August 6, the last time I purged data from the table. My code attempts to update once per hour and retries every 5 minutes if 2 fetch attempts fail. It has good and bad days, but its a constant issue that's been going on for a _long_ time.

```
 2022-08-06 |    12
 2022-08-07 |     9
 2022-08-08 |    14
 2022-08-09 |    14
 2022-08-10 |    15
 2022-08-11 |    10
 2022-08-12 |    17
 2022-08-13 |     7
 2022-08-14 |    20
 2022-08-15 |     7
 2022-08-16 |    15
 2022-08-17 |    10
 2022-08-18 |     8
 2022-08-19 |    12
 2022-08-20 |     6
 2022-08-21 |     5
 2022-08-22 |     7
 2022-08-23 |    14
 2022-08-24 |    10
 2022-08-25 |    15
 2022-08-26 |    10
 2022-08-27 |     6
 2022-08-28 |    20
 2022-08-29 |    13
```

**Comment by scadergit** [comment]
I hate to say it, but that's the exact question to ask NCO. It has been too long. Priorities get shifted (at times even away from the API completely), but the team is finally growing and hopefully can tackle.some key issues soon.

**Reply by dwhitemv25** [reply]
I can certainly turn up the heat on NCO a bit if it'd help. :)  My logging has all of the responses, and hopefully that is enough to concretely identify the issue. 

(Maybe not today, they're having a fun time in KC with dead & unsupported hardware. Takes me back...)

**Reply by jah1007** [reply]
This unfortunately is out of our control in NCO as our priorities are set by NWS HQ both in terms of development of the update as well as the implementation schedule for releases. There does seem to be some traction in hardening the timeline for repair as opposed to what we have had up to this point which is Very Soon™

**Comment by guarddog13** [comment]
I've been getting 500 error codes all day at 

https://api.weather.gov/gridpoints/DTX/42,58/forecast

Quite annoying as this breaks several home assistant automations.

**Comment by loukut** [comment]
I'm getting the 500 error codes also
https://api.weather.gov/gridpoints/MPX/87,80/forecast

correlationId: '298bf3d6', title: 'Unexpected Problem', type: 'https://api.weather.gov/problems/UnexpectedProblem', status: 500, detail: 'An unexpected problem has occurred.'

**Reply by ChrisGergler** [reply]
https://api.weather.gov/gridpoints/BOX/22,40/forecast/hourly Also gives similar Errors with 500 Error Code

**Comment by JJKraw** [comment]
Seems the KC outage problem is bad enough now that I can't even check the status at https://www.nco.ncep.noaa.gov/status/messages/

It times out.

**Reply by JJKraw** [reply]
Heh. Five minutes after writing that the message page is back.

**Reply by guarddog13** [reply]
Is that why their api is out in some areas?

**Reply by jah1007** [reply]
The KC datacenter outage did have a negative impact on the gridpoints endpoint of API. We implemented a workaround to restore the data feed.

**Reply by guarddog13** [reply]
@jah1007 A workaround? I'm still not getting any info on my integration. Is there something I need to do to implement the workaround?

**Reply by jah1007** [reply]
At last check there were still a handful of offices that had not sent in data since the service was restored and continue to throw 500 errors. That said you will continue to see sporadic 500 errors for the reasons listed above even after the remaining offices send in their data.

**Reply by guarddog13** [reply]
Thank you hopefully it's fixed soon.

**Reply by rbouchard13** [reply]
Is this for 500 or 503? The storms that we t through the other day have resulted in me getting 503 on almost every call

**Reply by rbouchard13** [reply]
I am in DTX gridpoint

**Reply by guarddog13** [reply]
I'm in DTX as well... I'm getting both 500 and 503s... mostly 503s.

**Reply by rbouchard13** [reply]
Check now! Looks like things are back up and running!@guarddog13

**Reply by guarddog13** [reply]
Thank you for telling me, thank goodness. I had automations erroring...like presence automations...my "weather ball" (really a light) was broken... my neighbors rely on this light for a quick forecast lol.

**Comment by krobine1** [comment]
Good day. I just found this today. I am also having the same issues. What's interesting though is that I have better luck with the same forecast locations if I try super early in the AM. But as it gets later in the day (MT), it seems to get worse and worse. I'm about to try the repeat query idea, but I'm writing an app that will run in both Desktop and Mobile devices so am afraid of performance if I have to do retries.  Makes me think the server processing these requests isn't supporting the number of requests being thrown at it?

**Reply by rbouchard13** [reply]
I implemented the retry on my app which is in both desktop and mobile web. its actually pretty quick! And you are correct that there are times of the day that seem worse. but recurrence can help with the 500

**Reply by krobine1** [reply]
rbouchard13:  Are you setting a timeout and doing the retry?  What's your timeout set to? How many retries? I'm getting closer to pushing my app up to production and want to make sure this is working better. I just implemented a retry of 3 times with a 300 millisecond wait time in between. It doesn't seem to help. Just wondering if there's a 'sweet spot' for this?

**Reply by rbouchard13** [reply]
@krobine1 My retry was based on the response. I run an async call to the api and then have condition based on the response status. If i dont get the 200, i just call it again

**Reply by krobine1** [reply]
How many times do you retry it?  In my code, if my response status shows a failure (ie. 500 message), I wait 300 milliseconds and then try again. I do this 3 times if it keeps failing. Then, I give up.

**Reply by rbouchard13** [reply]
couldn't you just make the function recursive until you get a 200? Let it try until it works! Granted not the best for network performance but =from what I have seen when you get the 500 you can usually get a valid response pretty quickly with refresh. UNLESS, that location is down for some reason

**Reply by krobine1** [reply]
Not with my app. It's a public-facing mobile/desktop 'big-usage' website (and mobile users may not have great coverage). Don't want users waiting for a response too long. If that were to happen, they wouldn't want to use the site (and I don't want to see that happen)...  Thanks for your responses though..

**Reply by rbouchard13** [reply]
what lat lng point are you calling?

**Reply by krobine1** [reply]
ha ha. All over the US. I have maybe 100 different locations I'm doing this on right now.  This is one example: https://api.weather.gov/gridpoints/VEF/110,98/forecast. Maybe in the future it could be on any lat/long in the US (including AK).

**Reply by rbouchard13** [reply]
The 500s are completely random, right? I don't think there would be a sweet spot for how long to wait that would change things. If they are refusing the call because of too much traffic it could be just as busy in 3 6 or 10 seconds. Wish there was a magic fix. But some of the end points are buggy

**Reply by krobine1** [reply]
Most of my locations are far away from cities. I've been playing around with the timeout and number of retries. I now have it set to wait 1 second in between attempts and try a total of 4 times. If it fails then, I'm out. I am having some decent success with that. I have those numbers configured so can easily change it later. Thanks again for your help..

**Reply by smoore4moma** [reply]
Well, it is free.  I would probably go with a paid api if it was mission-critical.

You might want to try adding random guids to tell the server that it is not a duplicate.  I seem to recall that it was needed but I don't see it in the documentation.

          'X-Correlation-Id': randomUUID(),
          'X-Request-Id': randomUUID(),

`    

      const url = '/gridpoints/OKX/33,37/forecast';
      console.log (url);

      var options = {
        'method': 'GET',
        'hostname': 'api.weather.gov',
        'path': url,
        'headers': {
          'Content-Type': 'application/geo+json',
          'Accept': '*/*' ,
          'Connection': 'keep-alive',
          'X-Correlation-Id': randomUUID(),
          'X-Request-Id': randomUUID(),
          'X-Server-Id': 'moma-nyc-weather',
          'User-Agent': 'moma-nyc-weather'
        },
        'maxRedirects': 20
      };`

**Comment by af6140** [comment]
This https://api.weather.gov/gridpoints/DTX/60,45/forecast does not work neither.

{
    "correlationId": "220808c",
    "title": "Unexpected Problem",
    "type": "https://api.weather.gov/problems/UnexpectedProblem",
    "status": 500,
    "detail": "An unexpected problem has occurred.",
    "instance": "https://api.weather.gov/requests/220808c"
}

**Comment by rr0ss0rr** [comment]
I’ve had pretty good success with a 2 second delay on failure with a max of 3 retries

**Comment by liamengland1** [comment]
`https://api.weather.gov/gridpoints/AER/126,236/forecast` seems to produce 500 errors more often than not.

**Comment by tk63135** [comment]
I submitted this as an operational issue today and thought I would share here in case it is helpful.  The requests below were working earlier today but they have stopped working.  There are other grid points that I am able to successfully access.

    https://api.weather.gov/gridpoints/BTV/126,66/forecast/hourly
    {
    "correlationId": "1ce89b44",
    "title": "Unexpected Problem",
    "type": "https://api.weather.gov/problems/UnexpectedProblem",
    "status": 500,
    "detail": "An unexpected problem has occurred.",
    "instance": "https://api.weather.gov/requests/1ce89b44"
    }

    https://api.weather.gov/gridpoints/BTV/105,59/forecast/hourly
    {
    "correlationId": "18cd6954",
    "title": "Unexpected Problem",
    "type": "https://api.weather.gov/problems/UnexpectedProblem",
    "status": 500,
    "detail": "An unexpected problem has occurred.",
    "instance": "https://api.weather.gov/requests/18cd6954"
    }

Any insights that can be provided are appreciated.

**Comment by forddavis** [comment]
I have been experiencing this as well, pretty consistently requesting two reports with a script in quick succession has been failing, even with an exponential backoff.

I'm not 100% sure this is the reason, but I believe not setting the headers tends to result in much more aggressive rate limiting:

[See Authentication spec](https://www.weather.gov/documentation/services-web-api#:~:text=Request%20new%20features-,Authentication,-A%20User%20Agent)
`User-Agent: (myweatherapp.com, contact@myweatherapp.com)`

After adding the headers it seems like I am very rarely if ever getting a 500 or even needing the backoff.

If this is truly the reason for the 500s, it would be much more user friendly if a 401 unauthorized response or 429 rate limiting response requesting a user agent was provided.

**Reply by strefethen** [reply]
FWIW, I've had a UA string set for a long time (years) calling this API and see failures all the time. I'm going to update to the above format and would love to see that make a difference though I'm not holding my breath. I'll avoid posting my UA given the advice from the docs but they do say the "string can be anything" and mine has included both my domain and contact email since I began calling the API.

The docs read:
```
This string can be anything, and the more unique to your application the less likely it will be affected by a security event. If you include
contact information (website or email), we can contact you if your string is associated to a security event. This will be replaced with an 
API key in the future.
```

You mention "headers" plural above, curious if it's only the User-Agent that you added?

**Comment by dwhitemv25** [comment]
There's been a couple of particularly bad days with 500 errors in the last week or so, enough that it generated traffic on the NCO-Ops master ticket. These are the error counts for my application for the last week (times as of midnight PDT):
```
  observed  | rc  | count 
------------+-----+-------
 2023-09-29 | 500 |     4
 2023-09-28 | 500 |    19
 2023-09-28 | 502 |     1
 2023-09-27 | 500 |     9
 2023-09-26 | 500 |    20
 2023-09-25 | 500 |     9
 2023-09-24 | 500 |     6
 2023-09-24 | 503 |     1
 2023-09-23 | 500 |     3
 2023-09-22 | 500 |     3
 2023-09-22 | 503 |     1
```
The 500s have never been due to rate limiting -- it is a resource limitation on NWS's end -- but caching does play a factor, and user-agent is used in determining if a request is from the same client or not. It was mentioned in the NCO ticket that there was a facility failover event and the errors peaked up just after that event. 

My use is either a frequent poll of /alerts/active for a couple of zones or an hourly poll of forecast data for one zone. My poller does a retry after about 5 seconds, which is usually enough for whatever backend issue caused the first 500 to clear. My average error count is about 6-7 500 errors per day, counting the retry, counting since Sept. 1, 2023.
